### PR TITLE
Add images/pull/tag integration test

### PR DESCRIPTION
**Pull request by jimmyxian**
Add images/pull/tag integration test
Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

**Comment by abronan** [comment]
`tag` test fails on my setup on `[[ "${lines[*]}" == *"tag_busybox"* ]]`. Not sure why though.. Is it the proper way to test on multiple lines output? Shouldn't we `grep tag_busybox` from the output instead?

/ping @aluzzardi

**Comment by vieux** [comment]
@jimmyxian @abronan we could `docker images tag_busybox`

File: test/integration/api.bats [hunk 1]
```diff
@@ -222,9 +312,30 @@ function teardown() {
...
+	[ "${#lines[@]}" -ge 2 ]
+	[[ "${lines[*]}" == *"busybox"* ]]
+	[[ "${lines[*]}" != *"tag_busybox"* ]]
+
+	# tag image
+	run docker_swarm tag busybox tag_busybox:test
+	[ "$status" -eq 0 ]
+
+	# verify
+	run docker_swarm images
```
Comment (vieux) [inline test/integration/api.bats]: please use `run docker_swarm images tag_busybox`

**Comment by abronan** [comment]
@vieux +1. Reduces the output and makes the test easier.

File: test/integration/api.bats [hunk 2]
```diff
@@ -222,9 +312,30 @@ function teardown() {
...
+
+	run docker_swarm pull busybox
+	[ "$status" -eq 0 ]
+
+	# make sure the image of busybox exists 
+	# the comming image of tag_busybox not exsit
+	run docker_swarm images
+	[ "$status" -eq 0 ]
+	[ "${#lines[@]}" -ge 2 ]
+	[[ "${lines[*]}" == *"busybox"* ]]
```
Comment (aluzzardi) [inline test/integration/api.bats]: Please use `$output` instead of `${lines[*]}`

**Comment by aluzzardi** [comment]
Related to #492

File: test/integration/api.bats [hunk 3]
```diff
@@ -68,9 +68,68 @@ function teardown() {
...
 }
 
-# FIXME
 @test "docker images" {
-	skip
+	start_docker 3
+	swarm_manage
+
+	# no image exist
+	run docker_swarm images
```
Comment (aluzzardi) [inline test/integration/api.bats]: I think here you could run `images -q` and simply check that the output has 0 lines

File: test/integration/api.bats [hunk 4]
```diff
@@ -68,9 +68,68 @@ function teardown() {
...
+
+	# no image exist
+	run docker_swarm images 
+	[ "$status" -eq 0 ]
+	[ "${#lines[@]}" -eq 1 ]
+	[[ "${lines[0]}" == *"REPOSITORY"* ]]
+
+	# make sure every node has no image
+	for((i=0;i<3;i++)); do
+		run docker_swarm images --filter node=node-$i
```
Comment (aluzzardi) [inline test/integration/api.bats]: `images -q`

File: test/integration/api.bats [hunk 5]
```diff
@@ -68,9 +68,68 @@ function teardown() {
...
+	# if there are 3 nodes, the output lines of "docker images" are greater or equal 4(1 header + 3 busybox:latest)
+	# so use -ge here, the following(pull/tag) is the same reason
+	[ "${#lines[@]}" -ge 4 ]
+	# Every line should contain "busybox" exclude the first head line 
+	for((i=1; i<${#lines[@]}; i++)); do
+		[[ "${lines[i]}" == *"busybox"* ]]
+	done
+}
+
+@test "docker images --filter node=<Node name>" {
```
Comment (aluzzardi) [inline test/integration/api.bats]: I think the two `images` tests can be combined

File: test/integration/api.bats [hunk 6]
```diff
@@ -162,9 +221,40 @@ function teardown() {
...
+	for host in ${HOSTS[@]}; do
+		run docker -H $host images -q
+		[ "$status" -eq 0 ]
+		[ "${#lines[@]}" -eq 0 ]
+	done
+
+	run docker_swarm pull busybox
+	[ "$status" -eq 0 ]
+
+	# swarm verify
```
Comment (aluzzardi) [inline test/integration/api.bats]: This is already checked by `docker images` - I think the node verify is enough

File: test/integration/api.bats (hunk 1 above)
Comment (jimmyxian) [inline test/integration/api.bats]: OK

File: test/integration/api.bats (hunk 3 above)
Comment (jimmyxian) [inline test/integration/api.bats]: OK

**Comment by jimmyxian** [comment]
@aluzzardi @abronan @vieux Thks
Aleady modifid by the comments

**Comment by abronan** [comment]
`tag` is still failing for me.

If I echo the `output` during the test I get:

```
REPOSITORY TAG IMAGE ID CREATED VIRTUAL SIZE
busybox latest 4986bf8c1536 3 months ago 2.433 MB
busybox latest 4986bf8c1536 3 months ago 2.433 MB
busybox latest 4986bf8c1536 3 months ago 2.433 MB
```

There is no tagged image so I wonder if there is actually a bug with `tag`.. Will check on a proper cluster.

**Comment by abronan** [comment]
Ok the test fails because it immediately issues a `docker images` right after the `docker tag` but the state is not refreshed so the output is the one I gave above: without the tagged image.

Just tested on a real cluster, if I `tag` an image and I do a `docker images` directly after, the `tag` does not appear in the output.

**Comment by aluzzardi** [comment]
It's probably the same issue we had with `build` and a few others, we need to force the refresh.

**Comment by abronan** [comment]
Ok this works great after rebasing off of Master. @jimmyxian could you please rebase so I can check again and confirm this works as expected. Thanks!

**Comment by jimmyxian** [comment]
@abronan My bad.Not rebase correctly yesterday
Done to Rebase Again. Thks

**Comment by abronan** [comment]
@jimmyxian No worries, this was me having a hard time to figure out where was the problem as I missed the PR with the force-refresh. Double-checked and works fine. Thanks again!

LGTM

/ping @aluzzardi

**Comment by aluzzardi** [comment]
LGTM

**Pull request merged**